### PR TITLE
dhclient: improve ifup code

### PR DESCRIPTION
**Pull request by rminnich**
The code should see if the link is up, and if so return immediately.
It should not return until the link is up.
30 seconds is about as long as one should try.

This will make it so you no longer have to hit ^C on the
first dhclient :-)

Signed-off-by: Ronald G. Minnich <rminnich@gmail.com>

### CODE REVIEW

File: cmds/dhclient/dhclient.go [hunk 1]
```diff
@@ -46,17 +47,27 @@ var (
...
+			debug("Link %v is up", ifname)
+			return iface, nil
+		}
 
+		if err := netlink.LinkSetUp(iface); err != nil {
+			return nil, fmt.Errorf("%v: %v can't make it up: %v", ifname, iface, err)
+		}
+		time.Sleep(1 * time.Second)
+	}
+	return nil, fmt.Errorf("Link %v still down after 30 seconds", ifname)
```
Comment (LaconicPneumonic) [inline cmds/dhclient/dhclient.go]: The "30" should probably be a format option for "linkUpAttempt". That way, it will be easier to change in the future.

File: cmds/dhclient/dhclient.go [hunk 2]
```diff
@@ -46,17 +47,27 @@ var (
...
 func ifup(ifname string) (netlink.Link, error) {
-	iface, err := netlink.LinkByName(ifname)
-	if err != nil {
-		return nil, fmt.Errorf("cannot get interface by name %v: %v", ifname, err)
-	}
-
-	if err := netlink.LinkSetUp(iface); err != nil {
-		return nil, fmt.Errorf("%v: %v can't make it up: %v", ifname, iface, err)
-	}
+	start := time.Now().Second()
```
Comment (ananion-cation) [inline cmds/dhclient/dhclient.go]: I tested the Now.Second method on my mac (go playground time does not work) and this only gives back a number from 0-59 (seconds in a minute) and then loops back. This is problematic for the for loop below if it started at 59 seconds and now will loop forever. An absolute value might help.

File: cmds/dhclient/dhclient.go [hunk 3]
```diff
@@ -46,17 +47,27 @@ var (
...
+			return nil, fmt.Errorf("cannot get interface by name %v: %v", ifname, err)
+		}
 
-	return iface, nil
+		if iface.Attrs().OperState == netlink.OperUp {
+			debug("Link %v is up", ifname)
+			return iface, nil
+		}
 
+		if err := netlink.LinkSetUp(iface); err != nil {
```
Comment (LaconicPneumonic) [inline cmds/dhclient/dhclient.go]: Should the order of "LinkSetUp" and "OperState == netlink.OperUp" be swapped? When you start dhclient for the first time, this loop will always go through a minimum of two times.

File: cmds/dhclient/dhclient.go (hunk 2 above)
Comment (ananion-cation) [inline cmds/dhclient/dhclient.go]: Using : tme := time.Now()
and 
for  int(time.Since(tme).Seconds()) < start {

}

worked for me: 

func main() {
	/*
	start := 59
	tmp := 0
	ntmp :=0
	for tmp - start < 30 {
		tmp = time.Now().Second()
		if ntmp != tmp {
			ntmp = tmp
			fmt.Printf("\n%d", ntmp)
		}
		
	}
	*/
	start := 30
	tme := time.Now()
	tmp := 0
	ntmp := 0
	for tmp < start {
		tmp := int(time.Since(tme).Seconds())
			if ntmp != tmp {
				ntmp = tmp
				fmt.Printf("\n%d", ntmp)
		}
	}
}

File: cmds/dhclient/dhclient.go (hunk 1 above)
Comment (rminnich) [inline cmds/dhclient/dhclient.go]: done

File: cmds/dhclient/dhclient.go [hunk 4]
```diff
@@ -33,6 +33,7 @@ const (
 	defaultIfname = "eth0"
 	// slop is the slop in our lease time.
 	slop = 10 * time.Second
+	linkUpAttempt = 30
 )
 
 var (
```
Comment (rminnich) [inline cmds/dhclient/dhclient.go]: done, it's not really part of this PR but it's an easy fix so why not :-)

File: cmds/dhclient/dhclient.go (hunk 2 above)
Comment (rminnich) [inline cmds/dhclient/dhclient.go]: done, thanks!

File: cmds/dhclient/dhclient.go (hunk 3 above)
Comment (rminnich) [inline cmds/dhclient/dhclient.go]: I think it's best to see if it's up before trying to do something to it, since some script or other action might have already set it up. I've observed we only go through the loop at 1.5 times (we bail on the second iteration when it is in OperUp state) so I think it's ok to do it this way. Thanks.

File: cmds/dhclient/dhclient.go [hunk 5]
```diff
@@ -30,33 +32,48 @@ import (
...
-	if err != nil {
-		return nil, fmt.Errorf("cannot get interface by name %v: %v", ifname, err)
-	}
-
-	if err := netlink.LinkSetUp(iface); err != nil {
-		return nil, fmt.Errorf("%v: %v can't make it up: %v", ifname, iface, err)
-	}
+	debug("Try bringing up %v", ifname)
+	start := time.Now()
+	for int(time.Since(start).Seconds()) < linkUpAttempt {
```
Comment (hugelgupf) [inline cmds/dhclient/dhclient.go]: make this

for time.Since(start) < linkUpAttempt

and 

const linkUpAttempt = 30 * time.Second

File: cmds/dhclient/dhclient.go [hunk 6]
```diff
@@ -247,12 +281,17 @@ func main() {
 			} else {
 				done <- dhclient6(iface, *renewals, timeout, *retry)
 			}
-		}(i)
+			debug("Done dhclient for %v", ifname)
+		}(i.Attrs().Name)
 	}
 
+	go func() {
+		wg.Wait()
```
Comment (hugelgupf) [inline cmds/dhclient/dhclient.go]: You don't need the WaitGroup because of the blocking, unbuffered done channel. I mean, this approach definitely works, too, but it's an additional goroutine and additional WaitGroup that you don't actually need.

Comment (hugelgupf) [inline cmds/dhclient/dhclient.go]: But you'll have to know how many goroutines exactly were launched. That's the only thing.

Comment (rminnich) [inline cmds/dhclient/dhclient.go]: I prefer the waitgroup, for that very reason (not having to know how many goroutines were launched). I don't see the cost of adding a goroutine as that big a deal -- heck, we're supposed to treat them as cheap. Since at some point I may want to do both v4 and v6 at the same time, which will mean an sync counter.

File: cmds/dhclient/dhclient.go [hunk 7]
```diff
@@ -30,33 +32,48 @@ import (
...
+		}
 
-	return iface, nil
+		if iface.Attrs().OperState == netlink.OperUp {
+			debug("Link %v is up", ifname)
+			return iface, nil
+		}
 
+		if err := netlink.LinkSetUp(iface); err != nil {
+			debug("Could not bring %v up: %v", ifname, err)
```
Comment (rminnich) [inline cmds/dhclient/dhclient.go]: yeah. I was in the middle of some desperate debugging and needed more info that I was getting but I'll remove some of these.

File: cmds/dhclient/dhclient.go (hunk 5 above)
Comment (rminnich) [inline cmds/dhclient/dhclient.go]: done

File: cmds/dhclient/dhclient.go (hunk 7 above)
Comment (rminnich) [inline cmds/dhclient/dhclient.go]: done

File: cmds/dhclient/dhclient.go (hunk 6 above)
Comment (rminnich) [inline cmds/dhclient/dhclient.go]: The waitgroup is actually a fairly standard way to solve this problem, I even see it used in places where the number of goroutines is known, since it continues to work should the code change in a way that removes that knowledge ...

**Comment by rminnich** [comment]
bump
